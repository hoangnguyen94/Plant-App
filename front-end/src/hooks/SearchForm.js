import SearchIcon from '@mui/icons-material/Search';
import { Typography } from '@mui/material';
import { useHistory } from 'react-router-dom';
import { v4 as uuid } from 'uuid';
import usePlantData from './usePlantData';
import React, { useState } from 'react';
import { TextField, Stack, Autocomplete } from '@mui/material';

const SearchForm = ({setSelectedPlantData}) => {
  const history = useHistory();
  const { searchPlant } = usePlantData();

  const [plants, setPlants] = useState([]);
  const [inputValue, setInputValue] = useState('');
  const [open, setOpen] = useState(false);
  
  const handleOnSubmit = (option) => {
    setSelectedPlantData( option );
    console.log( option.data );
    history.push( `/plants?name=${option.name}` );
    setOpen(false);
    setInputValue('');
  };

  const handleInputChange = async (e, name, reason) => {
    setInputValue(name);
    const results = await searchPlant(name);

    const plantsArray = Array.isArray(results) ? results : [results];
    const extractedPlants = plantsArray[0] && plantsArray[0].plants ? plantsArray[0].plants : [];
    let plantsWithId = extractedPlants.map((plant) => {
      return { ...plant, id: uuid() };
    });
    setPlants( plantsWithId );
    setOpen(!!name);
  };

  const filteredOptions = plants.filter((option) =>
    option.name && option.name.toLowerCase().includes(inputValue.toLowerCase())
  );

  return (
    <Stack sx={{ padding: 2, width: { xs: 200, sm: 400, md: 570 } }}>
      <Autocomplete
        clearOnBlur
        open={open}
        onOpen={() => {
          if (inputValue) {
            setOpen(true);
          }
        }}
        onClose={() => setOpen(false)}
        inputValue={inputValue}
        onInputChange={handleInputChange}
        popupIcon={<SearchIcon />}
        noOptionsText="We can't find your plant."
        getOptionLabel={(option) => `${option.name} ${option.sci_name}`}
        options={filteredOptions}
        renderInput={(params) => <TextField {...params} placeholder="Search..." />}
        renderOption={(props, option) => {
          return (
            <li
              {...props}
              key={option.id}
              onClick={() => {
                handleOnSubmit(option);
              }}
            >
              <Typography>
                {option.name} -
              </Typography>
              <Typography variant="subtitle2" sx={{ color: 'text.secondary' }}>
                <i>{option.sci_name}</i>
              </Typography>
            </li>
          );
        }}
      />
    </Stack>
  );
};

export default SearchForm;
