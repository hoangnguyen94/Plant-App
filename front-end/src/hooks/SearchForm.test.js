import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import SearchForm from './SearchForm';
import "mutationobserver-shim";

describe('SearchForm component', () => {
  test('renders SearchForm component correctly', () => {
    const setSelectedPlantData = jest.fn();

    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <SearchForm setSelectedPlantData={setSelectedPlantData} />
      </Router>
    );

    expect(container).toMatchSnapshot();
  });

  test('searches for plants and selects an option', async () => {
	const setSelectedPlantData = jest.fn();
	  setSelectedPlantData(
		  {
			  id: 1,
			  name: 'Name',
			  sci_name: 'plant Sci_name'
		  } )  
	const optionTextMatcher = (content, element) => {
		const hasDesiredText = Array.from(element.childNodes).some((node) => {
		  if (node.nodeType === Node.TEXT_NODE) {
			const text = node.textContent.trim();
			return text.includes('Plant Name');
		  }
		  return false;
		});
	  
		return hasDesiredText && element.tagName.toLowerCase() === 'li';
	  };
	  
    const history = createMemoryHistory();
    const { getByPlaceholderText, findByText } = render(
      <Router history={history}>
        <SearchForm setSelectedPlantData={setSelectedPlantData} />
      </Router>
	);
	
    const searchInput = getByPlaceholderText('Search...');
    fireEvent.change(searchInput, { target: { value: 'Plant Name' } });

    expect(setSelectedPlantData).toHaveBeenCalledWith({
      id: expect.any(Number),
      name: expect.any(String),
      sci_name: expect.any(String),
    });
  });
});
