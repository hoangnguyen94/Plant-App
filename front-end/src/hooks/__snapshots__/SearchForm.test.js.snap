// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm component renders SearchForm component correctly 1`] = `
<div>
  <div
    class="MuiStack-root css-1ut84p9-MuiStack-root"
  >
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-2"
            placeholder="Search..."
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
